examples/factory/src/lib.rs

```diff
@@ -0,0 +1,423 @@
+//! An example demonstrating how a factory pattern can be implemented with smart
+//! contracts on Concordium.
+//!
+//! > **Important Note:** The factory pattern is not generally idiomatic on
+//! > Concordium. In many cases, a more appropriate solution is to incorporate
+//! > the factory and products into a single smart contract instance, rather
+//! > than having a separate instance for each product.
```

Comment:
I've added a little more discussion. To me, maintaining an index of contracts is not an intrinsic part of the factory pattern - it is simply one smart contract that instantiates others. The fact that it can maintain a registry is incidental, and the main reason for doing it in this example is that there's no justification at all for using a factory if the factory doesn't have some additional functionality (on Concordium).

Reply:
I agree that usually, the factory does not assign a specific index to the contract registered. This is something we added in this example but it is not part of the factory pattern/standard.

What is important so is that the factory functions as a registry in many use cases for people to look up/resolve addresses off-chain. For example, if you want to trade tokens in the `ETH-CCD` token pair on a DEX in UNISWAP V2 design. You would look up if the `ETH-CCD` token pair is already registered in the factory to find the contract that you should interact with if you want to trade these tokens in a pool.
https://github.com/Uniswap/v2-core/blob/master/contracts/UniswapV2Factory.sol#L23